cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -699,13 +699,20 @@ predicate localExprFlow(Expr e1, Expr e2) { localFlow(exprNode(e1), exprNode(e2)
  */
 class BarrierGuard extends IRGuardCondition {
   /** Override this predicate to hold if this guard validates `instr` upon evaluating to `b`. */
-  abstract predicate checks(Instruction instr, boolean b);
+  predicate checksInstr(Instruction instr, boolean b) { none() }
+
+  /** Override this predicate to hold if this guard validates `expr` upon evaluating to `b`. */
+  predicate checks(Expr e, boolean b) { none() }
 
   /** Gets a node guarded by this guard. */
   final Node getAGuardedNode() {
     exists(ValueNumber value, boolean edge |
+      (
+        this.checksInstr(value.getAnInstruction(), edge)
+        or
+        this.checks(value.getAnInstruction().getConvertedResultExpression(), edge)
```

Comment:
Is there a reason this should be the _converted_ result expression? I don't know the exact behavior of the old barrier guard library, but will users not be surprised if they need to think of conversions when implementing `checks`?

Reply:
`getConvertedResultExpression` gets the expression that this instruction computes the result of, without adding or removing conversions, whereas `getUnconvertedResultExpression` strips all conversions. Using it here means that a a user specifying a conversion will work, where `getUnconvertedResultExpression` would require that the user's `check` implementation selects an unconverted expression.

Reply:
That indeed sounds like the right choice, then. Thanks for the explanation!